Track outdated nodes per bucket space

toregge opened the pull request:
@geirst, @vekterli: please review

Commit:
Pass distributor bucket space to pending bucket space db transition.

Commit:
Track outdated nodes per bucket space.
Use distributor bucket space to get distribution when detecting outdated nodes.

geirst reviewed (approved):
👍

vekterli commented on storage/src/vespa/storage/distributor/pendingclusterstate.h:
```diff
@@ -25,6 +25,8 @@ class DistributorBucketSpaceRepo;
 class PendingClusterState : public vespalib::XmlSerializable {
 public:
+    using OutdatedNodes = std::unordered_set<uint16_t>;
+    using OutdatedNodesMap = std::unordered_map<document::BucketSpace, OutdatedNodes, document::BucketSpace::hash>;
```
This typedef should ideally also be de-duplicated

vekterli commented on storage/src/vespa/storage/distributor/pending_bucket_space_db_transition.cpp:
```diff
@@ -11,20 +12,42 @@ LOG_SETUP(".pendingbucketspacedbtransition");
+                                                               bool distributionChanged,
+                                                               const OutdatedNodes &outdatedNodes,
                                                                std::shared_ptr<const ClusterInformation> clusterInfo,
+                                                               const lib::ClusterState &prevClusterState,
```
This constructor argument list size is growing outside my comfort zone 😨  We should consider if we can encapsulate this differently during subsequent refactoring.

vekterli reviewed (approved):
LGTM, see comments on typedef duplications

toregge commented on storage/src/vespa/storage/distributor/pending_bucket_space_db_transition.cpp:
```diff
@@ -11,20 +12,42 @@ LOG_SETUP(".pendingbucketspacedbtransition");
+
 PendingBucketSpaceDbTransition::PendingBucketSpaceDbTransition(const PendingClusterState &pendingClusterState,
+                                                               DistributorBucketSpace &distributorBucketSpace,
+                                                               int distributorIndex,
```
Distributor::getDistributorIndex() returns int. Should that be fixed too ?

Commit:
Add using statements for OutdatedNodes and OutdatedNodesMap in nested
namespace.
Reduce number of arguments to PendingBucketSpaceDbTransition constructor.

Commit:
Reference using statement in nested namespace.

vekterli reviewed (approved):
:+1:

geirst reviewed (approved):
👍
